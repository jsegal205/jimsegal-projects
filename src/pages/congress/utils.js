export const chamberTitles = (chamber) => {
  switch (chamber) {
    case "house":
      return "Representatives";
    case "senate":
      return "Senators";
    default:
      return "Members";
  }
};

export const individualChamberTitle = (chamber) => {
  const pluralTitle = chamberTitles(chamber);
  return pluralTitle.substring(0, pluralTitle.length - 1);
};

export const otherChamber = (chamber) =>
  chamber === "house" ? "senate" : "house";
