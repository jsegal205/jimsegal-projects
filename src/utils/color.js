const getRGB = (hex) => {
  return {
    r: parseInt(hex.slice(0, 2), 16),
    g: parseInt(hex.slice(2, 4), 16),
    b: parseInt(hex.slice(4, 6), 16),
  };
};

export const randomHexColor = () =>
  Math.floor(Math.random() * 16777215).toString(16);

export const complementaryColor = (hex) => {
  const padZero = (str) => ("00" + str).slice(-2);
  const invertColor = (digit) => padZero((255 - digit).toString(16));
  const { r, g, b } = getRGB(hex);

  return invertColor(r) + invertColor(g) + invertColor(b);
};

export const textColor = (hex) => {
  const { r, g, b } = getRGB(hex);
  const brightness = Math.round((r * 299 + g * 587 + b * 114) / 1000);

  return brightness > 125 ? "444444" : "efefef";
};
