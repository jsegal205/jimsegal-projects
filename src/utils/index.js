// export const apiUrlBase = "https://api.jimsegal.com";
export const apiUrlBase = "http://localhost:8001";

export const projectInfo = [
  {
    absolute: true,
    link: "https://jimsegal.com",
    subtitle: "OG site with other projects",
    title: "JimSegal.com",
  },
  {
    link: "/congress",
    subtitle: "Some fun data surrounding the current congressional sesssion",
    title: "Congress",
  },
  {
    link: "/isJimWearingShorts",
    subtitle: "A site to predict if Jim is wearing shorts or not",
    title: "Is Jim Wearing Shorts?",
  },
  {
    link: "mountaingoat",
    subtitle: "Digital game",
    title: "Mountain Goat Game",
  },
  {
    link: "/recipes",
    subtitle: "A bunch of recipes that I frequent and wanted to show off",
    title: "Recipes",
  },
  {
    link: "/spacex",
    subtitle: "A countdown timer until the next launch of a Space X rocket",
    title: "Space X",
  },
];
