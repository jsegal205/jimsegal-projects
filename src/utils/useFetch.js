import { useState, useEffect } from "react";
import axios from "axios";

// custom hook for performing GET request
const useFetch = (url, initialValue) => {
  const [data, setData] = useState(initialValue);
  const [loading, setLoading] = useState(true);
  useEffect(() => {
    const fetchData = async () => {
      try {
        setLoading(true);
        const response = await axios.get(url);
        setData(response.data);
      } catch ({ response }) {
        setData({ status: response.status });
      } finally {
        setLoading(false);
      }
    };
    fetchData();
  }, [url]);
  return { loading, data };
};

export default useFetch;
