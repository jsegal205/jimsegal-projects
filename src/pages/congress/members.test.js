import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";
import { render, fireEvent, getByTestId } from "@testing-library/react";
import { toMatchDiffSnapshot } from "snapshot-diff";

import { apiUrlBase } from "../../utils";
import useFetch from "../../utils/useFetch";
import CongressMembers from "./members";

expect.extend({ toMatchDiffSnapshot });

jest.mock("../../utils/useFetch", () => jest.fn());

describe("CongressMembers Component", () => {
  describe("when `useFetch` is awaiting promise to resolve", () => {
    it("displays `Loading...`", () => {
      const chamber = "doesNotExist";
      const props = {
        location: {
          pathname: `/congress/${chamber}/members`,
        },
      };

      useFetch.mockReturnValue({
        loading: true,
        data: {},
      });

      const tree = renderer.create(
        <Router>
          <CongressMembers {...props} />
        </Router>
      );
      expect(useFetch).toHaveBeenCalledWith(
        `${apiUrlBase}/congress/${chamber}/members`
      );
      expect(tree.toJSON()).toMatchSnapshot();
    });
  });

  describe("when `useFetch` returns data", () => {
    describe("with 404 status", () => {
      it("displays `Not Found`", () => {
        const chamber = "doesNotExist";
        const props = {
          location: {
            pathname: `/congress/${chamber}/members`,
          },
        };

        useFetch.mockReturnValue({
          loading: false,
          data: { status: 404 },
        });

        const tree = renderer.create(
          <Router>
            <CongressMembers {...props} />
          </Router>
        );
        expect(useFetch).toHaveBeenCalledWith(
          `${apiUrlBase}/congress/${chamber}/members`
        );
        expect(tree.toJSON()).toMatchSnapshot();
      });
    });

    describe("with 500 status", () => {
      it("displays error message", () => {
        const chamber = "house";
        const props = {
          location: {
            pathname: `/congress/${chamber}/members`,
          },
        };

        useFetch.mockReturnValue({
          loading: false,
          data: { error: "server likely on fire", status: 500 },
        });

        const tree = renderer.create(
          <Router>
            <CongressMembers {...props} />
          </Router>
        );
        expect(useFetch).toHaveBeenCalledWith(
          `${apiUrlBase}/congress/${chamber}/members`
        );
        expect(tree.toJSON()).toMatchSnapshot();
      });
    });

    describe("with 200 status", () => {
      it("displays data", () => {
        const chamber = "house";
        const props = {
          location: {
            pathname: `/congress/${chamber}/members`,
          },
        };

        useFetch.mockReturnValue({
          loading: false,
          data: {
            house: [
              {
                id: 1,
                first_name: "first",
                last_name: "last",
                party: "party",
                state: "state",
              },
            ],
            status: 200,
          },
        });

        const tree = renderer.create(
          <Router>
            <CongressMembers {...props} />
          </Router>
        );
        expect(useFetch).toHaveBeenCalledWith(
          `${apiUrlBase}/congress/${chamber}/members`
        );
        expect(tree.toJSON()).toMatchSnapshot();
      });
    });
  });

  describe("getChamber", () => {
    describe("when props contain state.chamber", () => {
      it("pulls chamber from state", () => {
        const props = {
          location: {
            state: {
              chamber: "state-chamber",
            },
            pathname: `/congress/house/members`,
          },
        };

        useFetch.mockReturnValue({
          loading: true,
          data: {},
        });

        renderer.create(
          <Router>
            <CongressMembers {...props} />
          </Router>
        );
        expect(useFetch).toHaveBeenCalledWith(
          `${apiUrlBase}/congress/state-chamber/members`
        );
      });
    });

    describe("when props do not contain state.chamber", () => {
      it("pulls slug from pathname", () => {
        const props = {
          location: {
            state: {
              "not-a-chamber": "state-chamber",
            },
            pathname: `/congress/house/members`,
          },
        };

        useFetch.mockReturnValue({
          loading: true,
          data: {},
        });

        renderer.create(
          <Router>
            <CongressMembers {...props} />
          </Router>
        );
        expect(useFetch).toHaveBeenCalledWith(
          `${apiUrlBase}/congress/house/members`
        );
      });
    });
  });

  describe("Search filtering", () => {
    const apiData = {
      house: [
        {
          id: 1,
          first_name: "abc",
          last_name: "efg",
          party: "party",
          state: "state",
        },
        {
          id: 2,
          first_name: "uxw",
          last_name: "xyz",
          party: "party",
          state: "state",
        },
      ],
    };

    const setup = () => {
      const props = {
        location: {
          pathname: `/congress/house/members`,
        },
      };

      useFetch.mockReturnValue({
        loading: false,
        data: apiData,
      });

      return render(
        <Router>
          <CongressMembers {...props} />
        </Router>
      );
    };

    it("defaults search input to empty string", () => {
      const { container } = setup();
      const membersFilter = getByTestId(container, "search-filter");
      const membersList = getByTestId(container, "members-list");
      const membersListItems = membersList.querySelectorAll("li");

      expect(membersFilter.value).toBe("");
      expect(membersListItems.length).toEqual(apiData["house"].length);
      expect(membersListItems[0].querySelector("a").innerHTML).toContain(
        apiData["house"][0].first_name
      );
      expect(membersListItems[1].querySelector("a").innerHTML).toContain(
        apiData["house"][1].first_name
      );
    });

    describe("when input matches subset of the data", () => {
      it("only shows data to match first name", () => {
        const { container, asFragment } = setup();

        const initialRender = asFragment();
        const memberFilter = getByTestId(container, "search-filter");
        fireEvent.change(memberFilter, { target: { value: "abc" } });

        expect(initialRender).toMatchDiffSnapshot(asFragment());
      });

      it("only shows data to match last name", () => {
        const { container, asFragment } = setup();

        const initialRender = asFragment();
        const memberFilter = getByTestId(container, "search-filter");
        fireEvent.change(memberFilter, { target: { value: "xyz" } });

        expect(initialRender).toMatchDiffSnapshot(asFragment());
      });
    });

    describe("when input matches none of the data", () => {
      it("only shows data to match", () => {
        const { container, asFragment } = setup();

        const initialRender = asFragment();
        const memberFilter = getByTestId(container, "search-filter");
        fireEvent.change(memberFilter, { target: { value: "mno" } });

        expect(initialRender).toMatchDiffSnapshot(asFragment());
      });
    });

    describe("when reset clicked", () => {
      it("resets search input value", () => {
        const { container } = setup();

        const recipeFilter = getByTestId(container, "search-filter");
        fireEvent.change(recipeFilter, { target: { value: "a" } });

        expect(recipeFilter.value).toBe("a");

        const recipeFilterReset = getByTestId(container, "search-filter-reset");

        fireEvent.click(recipeFilterReset);
        expect(recipeFilter.value).toBe("");
      });
    });
  });
});
