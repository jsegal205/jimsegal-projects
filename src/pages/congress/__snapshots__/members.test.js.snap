// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CongressMembers Component Search filtering when input matches none of the data only shows data to match 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -29,45 +29,26 @@
        <input
          class=\\"search-filter-input\\"
          data-testid=\\"search-filter\\"
          id=\\"search-filter\\"
          type=\\"text\\"
-         value=\\"\\"
+         value=\\"mno\\"
        />
-     </section>
-     <ul
-       class=\\"congress-members\\"
-       data-testid=\\"members-list\\"
-     >
-       <li>
-         <a
-           class=\\"congress-members-card\\"
-           href=\\"/congress/house/member/1\\"
+       <button
+         class=\\"search-filter-reset\\"
+         data-testid=\\"search-filter-reset\\"
        >
-           <img
-             alt=\\"Professional headshot of abc efg\\"
-             src=\\"https://www.congress.gov/img/member/1.jpg\\"
-           />
-           <label>
-             abc efg
-           </label>
-           Democrat - state
-         </a>
-       </li>
-       <li>
-         <a
-           class=\\"congress-members-card\\"
-           href=\\"/congress/house/member/2\\"
+         Reset
+       </button>
+     </section>
+     <div
+       class=\\"members-no-results\\"
      >
-           <img
-             alt=\\"Professional headshot of uxw xyz\\"
-             src=\\"https://www.congress.gov/img/member/2.jpg\\"
-           />
-           <label>
-             uxw xyz
-           </label>
-           Republican - state
-         </a>
-       </li>
-     </ul>
+       No results for 
+       <strong>
+         <em>
+           mno
+         </em>
+       </strong>
+     </div>
    </section>
  </DocumentFragment>"
`;

exports[`CongressMembers Component Search filtering when input matches subset of the data only shows data to match first name 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -29,12 +29,18 @@
        <input
          class=\\"search-filter-input\\"
          data-testid=\\"search-filter\\"
          id=\\"search-filter\\"
          type=\\"text\\"
-         value=\\"\\"
+         value=\\"abc\\"
        />
+       <button
+         class=\\"search-filter-reset\\"
+         data-testid=\\"search-filter-reset\\"
+       >
+         Reset
+       </button>
      </section>
      <ul
        class=\\"congress-members\\"
        data-testid=\\"members-list\\"
      >
@@ -49,25 +55,10 @@
            />
            <label>
              abc efg
            </label>
            Democrat - state
-         </a>
-       </li>
-       <li>
-         <a
-           class=\\"congress-members-card\\"
-           href=\\"/congress/house/member/2\\"
-         >
-           <img
-             alt=\\"Professional headshot of uxw xyz\\"
-             src=\\"https://www.congress.gov/img/member/2.jpg\\"
-           />
-           <label>
-             uxw xyz
-           </label>
-           Republican - state
          </a>
        </li>
      </ul>
    </section>
  </DocumentFragment>"
`;

exports[`CongressMembers Component Search filtering when input matches subset of the data only shows data to match last name 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -29,32 +29,23 @@
        <input
          class=\\"search-filter-input\\"
          data-testid=\\"search-filter\\"
          id=\\"search-filter\\"
          type=\\"text\\"
-         value=\\"\\"
+         value=\\"xyz\\"
        />
+       <button
+         class=\\"search-filter-reset\\"
+         data-testid=\\"search-filter-reset\\"
+       >
+         Reset
+       </button>
      </section>
      <ul
        class=\\"congress-members\\"
        data-testid=\\"members-list\\"
      >
-       <li>
-         <a
-           class=\\"congress-members-card\\"
-           href=\\"/congress/house/member/1\\"
-         >
-           <img
-             alt=\\"Professional headshot of abc efg\\"
-             src=\\"https://www.congress.gov/img/member/1.jpg\\"
-           />
-           <label>
-             abc efg
-           </label>
-           Democrat - state
-         </a>
-       </li>
        <li>
          <a
            class=\\"congress-members-card\\"
            href=\\"/congress/house/member/2\\"
          >"
`;

exports[`CongressMembers Component when \`useFetch\` is awaiting promise to resolve displays \`Loading...\` 1`] = `
<section
  className="loader"
>
  <div
    className="loader-wheel"
  />
  <div
    className="loader-text"
  >
    Loading
  </div>
</section>
`;

exports[`CongressMembers Component when \`useFetch\` returns data with 200 status displays data 1`] = `
<section>
  <a
    className="chamber-members"
    href="/congress"
    onClick={[Function]}
  >
    All Congress Data
  </a>
  <h2
    className="congress-header"
  >
    Current 
    Representatives
  </h2>
  <a
    className="chamber-members"
    href="/congress/senate/members"
    onClick={[Function]}
  >
    View Current 
    Senators
  </a>
  <hr />
  <section
    className="search-filter-container"
  >
    <label
      htmlFor="search-filter"
    >
      Search
    </label>
    <input
      className="search-filter-input"
      data-testid="search-filter"
      id="search-filter"
      onChange={[Function]}
      type="text"
      value=""
    />
    
  </section>
  <div
    className="members-no-results"
  >
    No results for
     
    <strong>
      <em>
        
      </em>
    </strong>
  </div>
</section>
`;

exports[`CongressMembers Component when \`useFetch\` returns data with 404 status displays \`Not Found\` 1`] = `
Array [
  <h2>
    Sorry, looks like I couldn't find what you were looking for.
  </h2>,
  <a
    href="/"
    onClick={[Function]}
  >
    Click to return somewhere familiar
  </a>,
]
`;

exports[`CongressMembers Component when \`useFetch\` returns data with 500 status displays error message 1`] = `
<section>
  <h3>
    Whoops! There was a problem loading
     
    Congress Memebers data.
  </h3>
  <p>
    Please reload browser to try again in a little bit.
  </p>
</section>
`;
